lgc/builder/BuilderImpl.cpp

```diff
@@ -74,6 +74,161 @@ Value *BuilderImplBase::CreateDotProduct(Value *const vector1, Value *const vect
+  // 2. The factor inputs must be <2 x i16> or <N x i8> (N <= 4) or <N x i4> (N <= 8)
+  const auto &supportIntegerDotFlag = getPipelineState()->getTargetInfo().getGpuProperty().supportIntegerDotFlag;
+  const bool isSupportCompBitwidth = (supportIntegerDotFlag.compBitwidth16 && compBitWidth == 16) ||
+                                     (supportIntegerDotFlag.compBitwidth8 && compBitWidth == 8) ||
+                                     (supportIntegerDotFlag.compBitwidth4 && compBitWidth == 4);
+  const bool isSupportSignedness =
+      isMixedSigned ? supportIntegerDotFlag.diffSignedness : supportIntegerDotFlag.sameSignedness;
+  const bool isDot2 = (compCount == 2 && compBitWidth == 16);
+  const bool isDot4 = (compCount <= 4 && compBitWidth == 8);
+  const bool isDot8 = (compCount <= 8 && compBitWidth == 4);
+  const bool hasNativeIntrinsic = isSupportCompBitwidth && isSupportSignedness && (isDot2 || isDot4 || isDot8);
+
+  // Create a lamda to clamp the input as expected
+  auto doClamp = [&](Value *input, Type *expectedTy) {
+    // Saturating
+    const unsigned resultWidth = expectedTy->getScalarSizeInBits();
+    // Compute the clamp range based on the
+    unsigned long long unsignedMax = (2ULL << (resultWidth - 1)) - 1;
+    long long signedMax = unsignedMax >> 1;
+    long long signedMin = -1LL - signedMax;
+
+    Value *minimum = nullptr, *maximum = nullptr;
+    Value *isUnderflow = nullptr, *isOverflow = nullptr;
+    if (isSigned) {
+      input = CreateSExt(input, getInt64Ty());
+      minimum = ConstantInt::getSigned(getInt64Ty(), signedMin);
+      maximum = ConstantInt::getSigned(getInt64Ty(), signedMax);
+      isUnderflow = CreateICmpSLT(input, minimum);
+      isOverflow = CreateICmpSGT(input, maximum);
+    } else {
+      input = CreateZExt(input, getInt64Ty());
+      minimum = getInt64(0);
+      maximum = getInt64(unsignedMax);
+      isUnderflow = CreateICmpULT(input, minimum);
+      isOverflow = CreateICmpUGT(input, maximum);
+    }
+    input = CreateSelect(isUnderflow, minimum, input);
+    input = CreateSelect(isOverflow, maximum, input);
+    return CreateTrunc(input, expectedTy);
+  };
+
+  auto input1 = vector1;
+  auto input2 = vector2;
+  Value *result = nullptr;
+  if (hasNativeIntrinsic) {
+    int supportedN = InvalidValue;
+    int intrinsic = InvalidValue;
+    if (isDot2) {
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot2 : Intrinsic::amdgcn_udot2;
+      supportedN = 2;
+    } else if (isDot4) {
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot4 : Intrinsic::amdgcn_udot4;
+      supportedN = 4;
+    } else {
+      assert(isDot8);
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot8 : Intrinsic::amdgcn_udot8;
+      supportedN = 8;
+    }
+    assert(intrinsic != InvalidValue);
+    // Do null-extension
+    SmallVector<int> shuffleMask;
+    for (int i = 0; i < supportedN; ++i)
+      shuffleMask.push_back(i);
```

Comment:
Consider the case where `compCount == 2` but `supportedN == 8`. The shuffleMask will go out of bounds. I'd suggest:
```suggestion
    for (int i = 0; i < supportedN; ++i)
      shuffleMask.push_back(std::min(i, compCount));
```

Reply:
Will fix.